### PR TITLE
[ECR-1495] CryptoCurrency demo transactions processing switched to bi…

### PR DESCRIPTION
**Pull request by DENMROOT**
…nary message format.

## Overview

CryptoCurrency demo backend transaction processing switched to binary message format.

---
See: https://jira.bf.local/browse/ECR-1495

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] The [continuous integration build](https://www.travis-ci.org/exonum/exonum-java-binding) passes

### CODE REVIEW

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxTest.java [hunk 1]
```diff
@@ -40,18 +41,15 @@
 import com.exonum.binding.storage.indices.MapIndex;
 import com.exonum.binding.test.RequiresNativeLibrary;
 import com.exonum.binding.util.LibraryLoader;
-import com.google.protobuf.ByteString;
 import nl.jqno.equalsverifier.EqualsVerifier;
 import org.junit.jupiter.api.Test;
 
-class CreateWalletTxTest {
+public class CreateWalletTxTest {
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxTest.java]: Junit 5 should work with package-private tests

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java [hunk 2]
```diff
@@ -16,28 +16,31 @@
...
 import static java.net.HttpURLConnection.HTTP_BAD_REQUEST;
 import static java.net.HttpURLConnection.HTTP_INTERNAL_ERROR;
 import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
 import static java.net.HttpURLConnection.HTTP_OK;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
```
[the remark this answers is not in the document]
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: No, it will just show some unexpected error (NPE, null)

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+   * Creates new unsigned binary create wallet message using provided key pair and provided balance.
+   */
+  static BinaryMessage createUnsignedMessage(PublicKey ownerKey, long initialBalance) {
+    return newCryptocurrencyTransactionBuilder(CreateWalletTx.ID)
+        .setBody(TxMessagesProtos.CreateWalletTx.newBuilder()
+            .setOwnerPublicKey(ByteString.copyFrom(ownerKey.toBytes()))
+            .setInitialBalance(initialBalance)
+            .build()
+            .toByteArray())
+        .setSignature(new byte[Message.SIGNATURE_SIZE])
```
[the remark this answers is not in the document]
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java]: Don't know,  it's just refactoring.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxTest.java (hunk 1 above)
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxTest.java]: You right. it's redundant

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java [hunk 4]
```diff
@@ -104,23 +104,21 @@ void tearDown(VertxTestContext context) {
...
+    BinaryMessage message = getBinaryMessage();
+
     String messageHash = "1234";
-    BinaryMessage message = mock(BinaryMessage.class);
     Transaction transaction = mock(Transaction.class);
 
-    when(jsonBinaryMessageConverter.toMessage(messageJson))
-        .thenReturn(message);
-    when(service.convertToTransaction(message))
+    when(service.convertToTransaction(any(BinaryMessage.class)))
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: Can we use `eq(message)` here?

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java [hunk 5]
```diff
@@ -218,6 +230,11 @@ void getWalletUsingInvalidKey(VertxTestContext context) {
         })));
   }
 
+  private BinaryMessage getBinaryMessage() {
+    KeyPair keyPair = CRYPTO_FUNCTION.generateKeyPair();
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: Why do we do this (generate a key pair and sign a message) if we don't check the signature anywhere?

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java]: No, it does not — the builder sets an empty signature by default.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import static com.exonum.binding.cryptocurrency.CryptocurrencyServiceImpl.CRYPTO_FUNCTION;
+import static com.exonum.binding.cryptocurrency.transactions.CryptocurrencyTransactionTemplate.newCryptocurrencyTransactionBuilder;
+
+import com.exonum.binding.crypto.KeyPair;
+import com.exonum.binding.crypto.PublicKey;
+import com.exonum.binding.messages.BinaryMessage;
+import com.exonum.binding.messages.Message;
+import com.google.protobuf.ByteString;
+
+public class CreateWalletTxBase {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java]: The name is confusing — shan't it be `CreateWalletMessages` or `CreateWalletTransactionMessages`?

It's likely this class doesn't have to be extracted — see above.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxTest.java [hunk 7]
```diff
@@ -237,16 +219,10 @@ void verifyEquals() {
         .verify();
   }
 
-  private static TransferTx withMockMessage(long seed, PublicKey senderId, PublicKey recipientId,
-                                            long amount) {
+  private TransferTx withMockMessage(long seed, PublicKey senderId, PublicKey recipientId,
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxTest.java]: Why this change — this method does not use any instance variables?

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxBase.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import com.exonum.binding.crypto.PublicKey;
+import com.exonum.binding.cryptocurrency.CryptocurrencySchema;
+import com.exonum.binding.cryptocurrency.Wallet;
+import com.exonum.binding.messages.BinaryMessage;
+import com.exonum.binding.storage.database.Fork;
+import com.exonum.binding.storage.indices.MapIndex;
+import com.google.protobuf.ByteString;
+
+
+public class TransferTxBase {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxBase.java]: Why this class is extracted? These methods were used by a specific subset of test methods in a single test class — does this PR change that? If not, I'd keep them.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxTest.java [hunk 9]
```diff
@@ -224,7 +205,8 @@ void info() {
     // Check the transaction parameters in JSON
     Gson gson = CryptocurrencyTransactionGson.instance();
 
-    Transaction txParameters = gson.fromJson(info, new TypeToken<TransferTx>() {}.getType());
+    Transaction txParameters = gson.fromJson(info, new TypeToken<TransferTx>() {
+    }.getType());
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxTest.java]: This doesn't seem to make it much clearer.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxBase.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+  /**
+   * Creates new signed binary transfer transaction message using provided keys and
+   * provided amount.
+   */
+  static BinaryMessage createSignedMessage(long seed, PublicKey senderId,
+      PrivateKey senderSecret,
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxBase.java]: If the indentation is changed, these two lines have to be joined.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java [hunk 11]
```diff
@@ -142,23 +140,37 @@ void submitValidTransaction(VertxTestContext context) {
...
   }
 
+  @Test
+  void submitTransactionOfIncorrectMessageSize(VertxTestContext context) {
+    BinaryMessage message = getBinaryMessage();
+    byte errorByte = 1;
+
+    post(ApiController.SUBMIT_TRANSACTION_PATH)
+        .sendBuffer(
+            Buffer.buffer(message.getSignedMessage().array()).appendByte(errorByte),
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: This method does not seem to test what it says it tests. As we use a mock of a service, `service.convertTransaction(message)` in the handler will return null (which is prohibited by the interface specification) — the additional byte won't change anything. I'll make this test fail _properly_ first to ensure that we get a correct error.

I suspect there either happens an NPE in the handler or a Mockito exception (because strict stubs are enabled).

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java [hunk 12]
```diff
@@ -218,6 +230,11 @@ void getWalletUsingInvalidKey(VertxTestContext context) {
         })));
   }
 
+  private BinaryMessage getBinaryMessage() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: `createTestMessage`?

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java [hunk 13]
```diff
@@ -218,6 +230,11 @@ void getWalletUsingInvalidKey(VertxTestContext context) {
...
+    KeyPair keyPair = CRYPTO_FUNCTION.generateKeyPair();
+    return createSignedMessage(keyPair);
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: As the handler doesn't care about the message, any can be passed, therefore, it might be sensible to just inline:
```java
newCryptocurrencyTransactionBuilder(CreateWalletTx.ID)
        .buildRaw();
```

A message with an empty payload will be produced ­— and that’s OK here.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java (hunk 4 above)
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: I think no, our test message and ApiController deserialized messages are not the same.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java (hunk 5 above)
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: It's redundant could be simplified.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java (hunk 6 above)
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTxBase.java]: Could be renamed, i just want to extract common functionality to some helper classes. Don't like when it's borrowed in tests.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxTest.java (hunk 9 above)
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/TransferTxTest.java]: it's my autoformat settings - sorry

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java (hunk 4 above)
Comment (dmitry-timofeev) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: I see, MessageReader does not implement equals, and as the whole hierarchy of Messages is about to be replaced with another one (hopefully, a simpler one), it doesn't make sense to implement e/h at the moment.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java (hunk 11 above)
Comment (dmitry-timofeev) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/ApiControllerTest.java]: I see, it tests invariants internal to _any_ message, but not to a message of a _particular_ transaction (that first invariant is 'payload_size' having a correct value). Do you think anything could be clarified to avoid such confusion?

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+   * Creates new signed binary create wallet message using provided key pair.
+   */
+  public static BinaryMessage createSignedMessage(KeyPair ownerKeyPair) {
+    BinaryMessage unsignedMessage = createUnsignedMessage(ownerKeyPair.getPublicKey(),
+        DEFAULT_BALANCE);
+    return unsignedMessage.sign(CRYPTO_FUNCTION, ownerKeyPair.getPrivateKey());
+  }
+
+  /**
+   * Creates new unsigned binary create wallet message using provided key pair and default balance.
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java]: the provided owner key

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+   */
+  public static BinaryMessage createUnsignedMessage(PublicKey ownerKey) {
+    return createUnsignedMessage(ownerKey, DEFAULT_BALANCE);
+  }
+
+  /**
+   * Creates new unsigned binary create wallet message using provided key pair and provided balance.
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java]: the provided owner key

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+import static com.exonum.binding.cryptocurrency.CryptocurrencyServiceImpl.CRYPTO_FUNCTION;
+import static com.exonum.binding.cryptocurrency.transactions.CryptocurrencyTransactionTemplate.newCryptocurrencyTransactionBuilder;
+
+import com.exonum.binding.crypto.KeyPair;
+import com.exonum.binding.crypto.PublicKey;
+import com.exonum.binding.messages.BinaryMessage;
+import com.exonum.binding.messages.Message;
+import com.google.protobuf.ByteString;
+
+public class CreateWalletTransactionUtils {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java]: I'd use the lowest needed access level for these utility classes and their methods as they are used in particular classes only (private or package-private); also, since they are utility classes, I'd forbid instantiation. This will make code suggestions as precise as they used to be and will help avoid confusion about this class purpose (as internal utility class).

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateTransferTransactionUtils.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+import com.google.protobuf.ByteString;
+
+
+public class CreateTransferTransactionUtils {
+
+  /**
+   * Creates new signed binary transfer transaction message using provided keys and
+   * provided amount.
+   */
+  public static BinaryMessage createSignedMessage(long seed, PublicKey senderId,
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateTransferTransactionUtils.java]: While we are at here, we unpack a `KeyPair` at call-site when we can do that here, reducing the number of arguments.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java (hunk 16 above)
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java]: Ok.

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
+package com.exonum.binding.cryptocurrency.transactions;
```
Comment (MakarovS) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java]: No copyright in this file

File: exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java [hunk 19]
```diff
@@ -9,30 +9,35 @@
 import com.exonum.binding.messages.Message;
 import com.google.protobuf.ByteString;
 
-public class CreateWalletTransactionUtils {
+class CreateWalletTransactionUtils {
 
-  public static final long DEFAULT_BALANCE = 100L;
+  static final long DEFAULT_BALANCE = 100L;
+
+  private CreateWalletTransactionUtils() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/src/test/java/com/exonum/binding/cryptocurrency/transactions/CreateWalletTransactionUtils.java]: Nit: usually put such things at the end as the least interesting in a class.

**Comment by dmitry-timofeev** [comment]
Merged into a separate branch to update the light-client later.
